File: go-leetcode/backtracking/77.组合.go
package leetcode

func combine(n int, k int) [][]int {
	result := [][]int{}
	path := []int{}

	var backtrack func(start int)
	backtrack = func(start int) {
		if len(path) == k {
			// 深拷贝路径
			temp := make([]int, k)
			copy(temp, path)
			result = append(result, temp)
			return
		}
		// 剪枝：最多从 start 到 n-(k-len(path))+1
		for i := start; i <= n-(k-len(path))+1; i++ {
			path = append(path, i)
			backtrack(i + 1)
			path = path[:len(path)-1]
		}
	}

	backtrack(1)
	return result
}

File: python-leetcode/backtracking/77.组合.py
from typing import List


class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        res: List[List[int]] = []
        path: List[int] = []

        def backtrack(start: int):
            if len(path) == k:
                res.append(path[:])
                return

            # 剪枝：最多从 start 到 n - (k - len(path)) + 1
            for i in range(start, n - (k - len(path)) + 2):
                path.append(i)
                backtrack(i + 1)
                path.pop()

        backtrack(1)
        return res
